distance-method/distance_method.py

```diff
@@ -0,0 +1,981 @@
+    output, error = '', ''
+
+    def __init__(self, command):
+        if isinstance(command, basestring):
+            command = shlex.split(command)
+        self.command = command
+
+    def run(self, timeout=None, **kwargs):
+        """ Run a command then return: (status, output, error). """
+        def target(**kwargs):
+            try:
+                self.process = subprocess.Popen(self.command, **kwargs)
+                self.output, self.error = self.process.communicate()
+                self.status = self.process.returncode
+            except:
+                self.error = traceback.format_exc()
+                self.status = -1
+        # default stdout and stderr
+        if 'stdout' not in kwargs:
+            kwargs['stdout'] = subprocess.PIPE
+        if 'stderr' not in kwargs:
+            kwargs['stderr'] = subprocess.PIPE
+        # thread
+        thread = threading.Thread(target=target, kwargs=kwargs)
+        thread.start()
+        thread.join(timeout)
+        if thread.is_alive():
+            self.process.terminate()
+            thread.join()
+        return self.status, self.output, self.error
+
+
+def hamming(str1, str2):
+    "Compute the Hamming distance between two strings"
+    assert len(str1) == len(str2)
+    ne = operator.ne
+    return sum(imap(ne, str1, str2))
+
+
+def preprocess_data(working_dir,
+                    target_proteomes_dir,
+                    verbose=False):
+    """ Map each gene to sudo name (ex. 1_1 for species 1, gene 1) for easier
+    output comparison and the 10 character name limitation in PHYLIP output.
+    This format is limited up to 9999 species and 99999 genes per species.
+
+    Parameters
+    ----------
+    working_dir:  string
+      path to working directory
+    target_proteomes_dir: string
+      path to directory holding proteomes for all target organisms
+    verbose: boolean, optional
+      output details about the running processes of this function
+
+    Returns
+    -------
+    gene_map: dictionary
+      "two-way" dictionary storing gene names as keys and their pseudo
+      names as values, and vica versa
+    ref_db: dictionary
+      dictionary storing FASTA label as key and sequence as value for the
+      reference databases
+    species: integer
+      the number of species in the reference databases
+    """
+    gene_map = {}
+    ref_db = {}
+    if verbose:
+        sys.stdout.write("Target organism\tNumber of genes\n")
+    # each file contains genes for species
+    species = 0
+    for _file in listdir(target_proteomes_dir):
+        if verbose:
+            sys.stdout.write("%s. %s\t" % (species+1, basename(_file)))
+        with open(join(target_proteomes_dir, _file), 'rb') as readfile:
+            gene = 0
+            for label, seq in parse_fasta(readfile):
+                label = label.split()[0]
+                ref_db[label] = seq
+                sudo_label = "%s_%s" % (species, gene)
+                gene_map[label] = sudo_label
```

Comment:
Will it be worth verifying that the name doesn't exist before adding it? Otherwise it will overwrite values silently.